libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/arrow/api/MapboxRouteArrowApi.kt

```diff
@@ -61,101 +70,194 @@ import com.mapbox.turf.TurfMeasurement
+        val arrowHeadFeatureCollection = getArrowHeadFeatureCollection()
+        return ArrowAddedValue(
+            shaftFeatureCollection,
+            arrowHeadFeatureCollection
+        )
     }
 
     /**
      * Calculates a maneuver arrow based on the route progress and returns a state that can
      * be used to render the arrow on the map.
      *
-     * @param routeProgress a route progress generated by the core navigation system.
+     * @param routeProgress a [RouteProgress] generated by the core navigation system.
      *
-     * @return the UpdateManeuverArrowState for rendering by the view.
+     * @return the [Expected<UpdateManeuverArrowValue, RouteArrowUpdateError>] for rendering by the view.
      */
-    fun updateUpcomingManeuverArrow(
-        routeProgress: RouteProgress
-    ): RouteArrowState.UpdateManeuverArrowState {
-        val invalidUpcomingStepPoints = (
-            routeProgress.upcomingStepPoints == null ||
-                routeProgress.upcomingStepPoints!!.size < RouteConstants.TWO_POINTS
+    fun addUpComingManeuverArrow(routeProgress: RouteProgress):
+        Expected<UpdateManeuverArrowValue, RouteArrowUpdateError> {
+            val invalidUpcomingStepPoints = (
+                routeProgress.upcomingStepPoints == null ||
+                    routeProgress.upcomingStepPoints!!.size < RouteConstants.TWO_POINTS
+                )
+            val invalidCurrentStepPoints = routeProgress.currentLegProgress == null ||
+                routeProgress.currentLegProgress!!.currentStepProgress == null ||
+                routeProgress.currentLegProgress!!.currentStepProgress!!.stepPoints == null ||
+                routeProgress.currentLegProgress!!.currentStepProgress!!.stepPoints!!.size <
+                RouteConstants.TWO_POINTS
+
+            val visibilityChanges = if (invalidUpcomingStepPoints || invalidCurrentStepPoints) {
+                getHideArrowModifications()
+            } else {
+                getShowArrowModifications()
+            }
+
+            removeArrow(maneuverPoints)
+            maneuverPoints = RouteArrowUtils.obtainArrowPointsFrom(routeProgress)
+            return when (val addArrowResult = addArrow(maneuverPoints)) {
+                is Expected.Failure -> {
+                    addArrowResult
+                }
+                is Expected.Success -> {
+                    Expected.Success(
+                        UpdateManeuverArrowValue(
+                            visibilityChanges,
+                            addArrowResult.value
+                                .arrowShaftFeatureCollection
+                                .features()
+                                ?.firstOrNull(),
+                            addArrowResult.value
+                                .arrowHeadFeatureCollection
+                                .features()
+                                ?.firstOrNull()
+                        )
+                    )
+                }
+            }
```

Comment:
Could we extract some of the logic into well-named `private` `fun`ctions?

Reply:
This is mostly a copy/paste from the 1.x implementation. I can separate some of it.